plutus-core/plutus-core/src/PlutusCore/TypeCheck/Internal.hs

```diff
@@ -82,19 +84,92 @@ functions that cannot fail look like this:
 -- ## Type definitions ##
 -- ######################
 
--- | Mapping from 'Builtin's to their 'Type's.
+-- | Mapping from 'Builtin's to their 'Normalized' 'Type's.
 newtype BuiltinTypes uni fun = BuiltinTypes
     { unBuiltinTypes :: Array fun (Dupable (Normalized (Type TyName uni ())))
     }
 
-type TyVarKinds = UniqueMap TypeUnique (Kind ())
-type VarTypes uni = UniqueMap TermUnique (Dupable (Normalized (Type TyName uni ())))
+type TyVarKinds = UniqueMap TypeUnique (Named (Kind ()))
+type VarTypes uni = UniqueMap TermUnique (Named (Dupable (Normalized (Type TyName uni ()))))
+
+-- | Decides what to do upon encountering a varible whose names doesn't match the name of the
+-- variable with the same unique that is currently in the scope. Consider for example this type:
+--
+--     \(a_0 :: *) (b_0 :: *) -> a_0
+--
+-- here @b_0@ shadows @a_0@ and so any variable having the @0@th unique within the body of the
+-- lambda references @b_0@, but we have @a_0@ there and so there's a name mismatch. Technically,
+-- it's not a type error to have a name mismatch, because uniques are the single source of truth,
+-- however a name mismatch is deeply suspicious and is likely to be caused by a bug somewhere.
```

Comment:
Has this actually happened?

Reply:
... and do you have any tests to check that it does catch the thing that's not supposed to happen? (I didn't see any, but maybe I didn't look closely enough.)

Reply:
> Has this actually happened?

[Yes](https://github.com/input-output-hk/plutus/pull/4559#issuecomment-1169915018).

> ... and do you have any tests to check that it does catch the thing that's not supposed to happen? (I didn't see any, but maybe I didn't look closely enough.)

I thought I'd added them! Stupid me, fixed now, thanks for catching.

Reply:
> [Yes](https://github.com/input-output-hk/plutus/pull/4559#issuecomment-1169915018).

Oh, it's that.  I saw the messages going past but didn't follow the details because it looked frightening. OK, it's very worthwhile to check for stuff like that.  It's kind of impressive how long it took for the issue to show up.